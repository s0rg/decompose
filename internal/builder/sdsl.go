//go:build !test

package builder

import (
	"errors"
	"fmt"
	"io"

	"github.com/s0rg/decompose/internal/node"
	sdsl "github.com/s0rg/decompose/internal/structurizr"
)

const systemName = "de-composed system"

var ErrDuplicate = errors.New("duplicate found")

type Structurizr struct {
	state *sdsl.System
}

func NewStructurizr() *Structurizr {
	return &Structurizr{
		state: sdsl.NewSystem(systemName),
	}
}

func (s *Structurizr) Name() string {
	return "structurizr-dsl"
}

func (s *Structurizr) AddNode(n *node.Node) error {
	cont, ok := s.state.AddContainer(n.ID, n.Name)
	if !ok {
		return fmt.Errorf("%w: %s", ErrDuplicate, n.Name)
	}

	cont.Technology = n.Image
	cont.Tags = make([]string, 0, len(n.Ports)+len(n.Networks))

	for _, p := range n.Ports {
		cont.Tags = append(cont.Tags, "listen:"+p.Label())
	}

	for _, n := range n.Networks {
		cont.Tags = append(cont.Tags, "net:"+n)
	}

	if n.IsExternal() {
		cont.Tags = append(cont.Tags, "external")
	}

	if n.Meta != nil {
		cont.Description = n.Meta.Info
		cont.Tags = append(cont.Tags, n.Meta.Tags...)
	}

	return nil
}

func (s *Structurizr) AddEdge(srcID, dstID string, port node.Port) {
	rel, ok := s.state.AddRelation(srcID, dstID)
	if !ok {
		return
	}

	rel.Tags = append(rel.Tags, port.Label())
}

func (s *Structurizr) Write(w io.Writer) {
	ws := sdsl.Workspace{
		System: s.state,
	}

	ws.Write(w)
}
