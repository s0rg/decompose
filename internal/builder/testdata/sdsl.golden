workspace {
	name "de-composed system"

	model {
		default = softwareSystem "default" {
		tags "3"
			3 = container "3" {
				description "info 3"
				technology "node-image"
				tags "3,listen:1/tcp,listen:2/tcp,net:test-net"
			}
		}
		c1 = softwareSystem "c1" {
		tags "1,2"
			1 = container "1" {
				description "info 1 \
docs-url \
repo-url"
				technology "node-image"
				tags "1,listen:1/tcp,listen:2/tcp,net:test-net"
			}
			2 = container "2" {
				description "info 2"
				technology "node-image"
				tags "2,listen:1/tcp,listen:2/tcp,net:test-net"
			}
		}
		c2 = softwareSystem "c2" {
		tags "ext2"
			ext2 = container "ext2" {
				tags "external,listen:2/tcp"
			}
		}

		c1 -> c2 {
			tags "0/"
		}
		c1 -> default {
			tags "0/"
		}
		default -> c2 {
			tags "0/"
		}
	}

	views {
		systemContext default "systemContext_default" {
			include *
			include c1
			include c2
			autoLayout
		}
		container default "container_default" {
			include *
			autoLayout
		}
		systemContext c1 "systemContext_c1" {
			include *
			autoLayout
		}
		container c1 "container_c1" {
			include *
			autoLayout
		}
		systemContext c2 "systemContext_c2" {
			include *
			autoLayout
		}
		container c2 "container_c2" {
			include *
			autoLayout
		}

		styles {
			element "Element" {
				metadata true
				description true
			}
		}
	}
}
